cli/lib.rs

```diff
@@ -371,12 +371,17 @@ fn run_script(flags: DenoFlags, argv: Vec<String>) {
     js_check(worker.execute("denoMain()"));
     debug!("main_module {}", main_module);
 
+    let mut worker_ = worker.clone();
+
     worker
       .execute_mod_async(&main_module, false)
       .and_then(move |()| {
         js_check(worker.execute("window.dispatchEvent(new Event('load'))"));
-        worker.then(|result| {
+        worker.then(move |result| {
           js_check(result);
+          js_check(
+            worker_.execute("window.dispatchEvent(new Event('unload'))"),
+          );
```

Comment:
Have you verified if async operations on unload would fire? Or would the Deno process just exit right away?

Reply:
That's a great question. I checked at MDN if `onunload` supports async functions but did not find any information.

I checked in Chrome with this example:
```js
async function wait(ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms);
  });
}

window.onunload = async function () { 
    console.log('before'); 
    await wait(1500); 
    console.log('after');
}
```
And it only prints `before` which suggests that promises are ignored. 

My implementation follows this behavior, but I'd prefer to get some second opinions because I believe that having async function as `unload` handler can be very useful. 

CC @ry @kitsonk @kevinkassimo

Reply:
Browsers specifically try to prevent scripts from delaying or blocking unload.

Notice that the difference between the `beforeunload` and `unload` events is that only the former is cancelable.

 - https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event
 - https://developer.mozilla.org/en-US/docs/Web/API/Window/unload_event

In practice, browsers also take other steps to prevent running excessive code during `unload`. I think it's reasonable for Deno to only support synchronous code in `unload`.

Reply:
I agree... We should maintain consistency with the web.  `beforeunload` should be cancelable, and allow the turn to complete, and drain the task queue before `unload` occurs.  `unload` should be the last turn, not executing any scheduled tasks.  This only deals with `unload` so it should just exit at the end of turn and not be cancelable.